src/Gui/Action.cpp

```diff
@@ -1243,4 +1350,206 @@ void WindowAction::addTo ( QWidget * w )
+    {
+        return QModelIndex();
+    }
+
+    virtual QVariant data(const QModelIndex & index, int role) const
+    {
+        if (index.row() < 0 || index.row() >= (int)_Commands.size())
+            return QVariant();
+
+        auto &info = _Commands[index.row()];
+
+        switch(role) {
+        case Qt::DisplayRole:
+        case Qt::EditRole:
+            if (info.text.isEmpty()) {
+#if QT_VERSION>=QT_VERSION_CHECK(5,2,0)
+                info.text = QString::fromLatin1("%2 (%1)").arg(
+                        QString::fromLatin1(info.cmd->getName()),
+                        qApp->translate(info.cmd->className(), info.cmd->getMenuText()));
+#else
+                info.text = qApp->translate(info.cmd->className(), info.cmd->getMenuText());
+#endif
+                info.text.replace(QLatin1Char('&'), QString());
+                if (info.text.isEmpty())
+                    info.text = QString::fromLatin1(info.cmd->getName());
+            }
+            return info.text;
+
+        case Qt::DecorationRole:
+            if (!info.iconChecked) {
+                info.iconChecked = true;
+                if(info.cmd->getPixmap())
+                    info.icon = BitmapFactory().iconFromTheme(info.cmd->getPixmap());
+            }
+            return info.icon;
+
+        case Qt::ToolTipRole:
+            if (info.tooltip.isEmpty()) {
+                info.tooltip = QString::fromLatin1("%1: %2").arg(
+                        QString::fromLatin1(info.cmd->getName()),
+                        qApp->translate(info.cmd->className(), info.cmd->getMenuText()));
+                QString tooltip = qApp->translate(info.cmd->className(), info.cmd->getToolTipText());
+                if (tooltip.size())
+                    info.tooltip += QString::fromLatin1("\n\n") + tooltip;
+            }
+            return info.tooltip;
+
+        case Qt::UserRole:
+            return QByteArray(info.cmd->getName());
+
+        default:
+            return QVariant();
+        }
+    }
+
+    virtual QModelIndex index(int row, int, const QModelIndex &) const
+    {
+        return this->createIndex(row, 0);
+    }
+
+    virtual int rowCount(const QModelIndex &) const
+    {
+        return (int)(_Commands.size());
+    }
+
+    virtual int columnCount(const QModelIndex &) const
+    {
+        return 1;
+    }
+};
+
+
+// --------------------------------------------------------------------
+
+CommandCompleter::CommandCompleter(QLineEdit *lineedit, QObject *parent)
+    : QCompleter(parent)
+{
+    this->setModel(new CommandModel(this));
+#if QT_VERSION>=QT_VERSION_CHECK(5,2,0)
+    this->setFilterMode(Qt::MatchContains);
+#endif
+    this->setCaseSensitivity(Qt::CaseInsensitive);
+    this->setCompletionMode(QCompleter::PopupCompletion);
+    this->setWidget(lineedit);
+    connect(lineedit, SIGNAL(textEdited(QString)), this, SLOT(onTextChanged(QString)));
+    connect(this, SIGNAL(activated(QModelIndex)), this, SLOT(onCommandActivated(QModelIndex)));

```

Comment:
Prefer new-style connections.

Reply:
Noted. I do use new connection now. These code were added before we fully switch to Qt5.